src/operations/client_bulk_write/command_builder.ts

```diff
@@ -0,0 +1,302 @@
+  ClientDeleteManyModel,
+  ClientDeleteOneModel,
+  ClientInsertOneModel,
+  ClientReplaceOneModel,
+  ClientUpdateManyModel,
+  ClientUpdateOneModel
+} from './common';
+
+/** @internal */
+export class ClientBulkWriteCommandBuilder {
+  models: AnyClientBulkWriteModel[];
+  options: ClientBulkWriteOptions;
+
+  /**
+   * Create the command builder.
+   * @param models - The client write models.
+   */
+  constructor(models: AnyClientBulkWriteModel[], options: ClientBulkWriteOptions) {
+    this.models = models;
+    this.options = options;
+  }
+
+  /**
+   * Gets the errorsOnly value for the command, which is the inverse of the
+   * user provided verboseResults option. Defaults to true.
+   */
+  get errorsOnly(): boolean {
+    if ('verboseResults' in this.options) {
+      return !this.options.verboseResults;
+    }
+    return true;
+  }
+
+  /**
+   * Build the bulk write commands from the models.
+   */
+  buildCommands(): Document[] {
+    // The base command.
+    const command: Document = {
+      bulkWrite: 1,
+      errorsOnly: this.errorsOnly,
+      ordered: this.options.ordered ?? true
+    };
+    // Add bypassDocumentValidation if it was present in the options.
+    if ('bypassDocumentValidation' in this.options) {
+      command.bypassDocumentValidation = this.options.bypassDocumentValidation;
+    }
+    // Add let if it was present in the options.
+    if ('let' in this.options) {
+      command.let = this.options.let;
+    }
+
+    // Iterate the models to build the ops and nsInfo fields.
+    const operations = [];
+    let currentNamespaceIndex = 0;
+    const namespaces = new Map<string, number>();
+    for (const model of this.models) {
+      const ns = model.namespace;
+      if (namespaces.has(ns)) {
+        operations.push(builderFor(model).buildOperation(namespaces.get(ns) as number));
+      } else {
+        namespaces.set(ns, currentNamespaceIndex);
+        operations.push(builderFor(model).buildOperation(currentNamespaceIndex));
+        currentNamespaceIndex++;
+      }
+    }
+
+    const nsInfo = Array.from(namespaces.keys()).map(ns => {
+      return { ns: ns };
+    });
+    command.ops = new DocumentSequence(operations);
+    command.nsInfo = new DocumentSequence(nsInfo);
+    return [command];
+  }
+}
+
+/** @internal */
+export interface OperationBuilder {
+  buildOperation(index: number): Document;
+}
+
+/**
+ * Builds insert one operations given the model.
+ * @internal
+ */
+export class InsertOneOperationBuilder implements OperationBuilder {
+  model: ClientInsertOneModel;
+
+  /**
+   * Instantiate the builder.
+   * @param model - The client insert one model.
+   */
+  constructor(model: ClientInsertOneModel) {
+    this.model = model;
+  }
+
+  /**
+   * Build the operation.
+   * @param index - The namespace index.
+   * @returns the operation.
+   */
+  buildOperation(index: number): Document {
+    const document: Document = {
+      insert: index,
+      document: this.model.document
```

Comment:
We likely need to autogenerate an object id using our `maybeAddIdToDocuments` helper (although, this method currently accepts a collection object - we'll have to change that :) ).

That also leads me to a second question - should we support CRUD-method specific options too?  For example - we support `pkFactory` and `forceServerObjectId` to insert operations.  We should likely support those for the new bulk write too.  Although pkFactory is tricky - because we usually get the pkFactory from the collection, which we don't have.

Reply:
edit: I found it in the spec, we do need to add _ids: https://github.com/mongodb/specifications/blob/master/source/crud/bulk-write.md#insert

Reply:
Note as I mentioned I am doing _id generation in the next ticket.